js/snake.js

```diff
@@ -1,305 +1,417 @@
+            this.gameOver();
+        }
+
+        // checks collision with food
+        else if (this.collides(this.food, this.snake.body[0])) {
+            this.field.score.set(this.snake.body.length - (this.snake.initLength));
+            this.placeFood();
+            //Speedup
+            if (Math.floor(this.tickrate / this.SPEEDUP) > this.SPEEDLIMIT) {
+                this.tickrate = Math.floor(this.tickrate / this.SPEEDUP);
 
-async function main(){
-    let snake;
-    field.init(document.body);
-    snake = new Snake(3);
-    document.addEventListener("keydown", this.keyDownHandler, false);
-    window.addEventListener("resize", field.resizeCanvas, false);
-    while(true){
-        snake.renderBody(field);
-        await sleep(field.tickrate);
-        snake.move(field.key);
+            }
+        }
+        else {
+            this.field.renderSegment(this.snake.body.pop(), this.BGCOLOR); //Canvas hinter snake mit Hintergrundfarbe überschreiben
+        }
     }
-}   
 
-function sleep(milliseconds) {  
-    return new Promise(resolve => setTimeout(resolve, milliseconds));
-}  
+    /**
+     * places Food on an empty field
+     */
+    placeFood() {
+        var free = this.notBody();
+        console.log(free)
+        this.food = free[Math.floor(Math.random() * free.length)];
+        this.field.renderSegment(this.food, this.DBIRED);
+    }
+
+    /**
+     * handles what happens at the end of a game
+     */
+    gameOver() {
+        console.log(this.snake.body.length + "-" + this.snake.initLength);
+        this.field.score.set(this.snake.body.length - (this.snake.initLength + 1));
+        if (this.field.score.score > this.field.score.highscore) {
+            alert("GAME OVER!!! \nScore: " + this.field.score.score + "\nNEW HIGHSCORE!!! \nPrevious Highscore: " + this.field.score.highscore);
+            window.localStorage.setItem("highscore", this.field.score.score);
+        }
+        else {
+            alert("GAME OVER!!! \nScore: " + this.field.score.score + "\nHighscore: " + this.field.score.highscore);
+        }
+        //Restart
+        var discard = this.snake.body.slice(this.snake.initLength, this.snake.body.length);
+        discard.forEach(segment => {
+            this.field.renderSegment(segment, this.BGCOLOR);
+        });
+        this.tickrate = this.INITSPEED;
+        this.snake.body = this.snake.body.slice(0, this.snake.initLength);
+        this.key = [];
+        this.field.score.set(0);
+        this.field.score.update();
+    }
 
+    /**
+     * evaluates if two Coord-objects refer to the same coordinates
+     * @param {Coord} obj1 
+     * @param {Coord} obj2 
+     * @returns {boolean}
+     */
+    collides(obj1, obj2) {
+        return obj1.x == obj2.x && obj1.y == obj2.y;
+    }
+
+    /**
+     * generates an array with every field the snake is NOT on
+     * @returns {Coord[]}
+     */
+    notBody() {
+        //generate all available fields
+        var free = [];
```

Comment:
Du könntest hier die Variable statt mit var einzuleiten auch mit "let" oder in diesem Fall, da du sie nicht neu zuweist, sondern nur den Inhalt veränderst, mit "const" initialisieren. Ist eines der neueren ES Features, welche eine Anwendung durchaus robuster machen. Ist aber optional :)